platforms/core-execution/workers/src/main/java/org/gradle/workers/internal/WorkerDaemonClientsManager.java

```diff
@@ -85,24 +80,42 @@ WorkerDaemonClient reserveIdleClient(DaemonForkOptions forkOptions, List<WorkerD
             Iterator<WorkerDaemonClient> it = clients.iterator();
             while (it.hasNext()) {
                 WorkerDaemonClient candidate = it.next();
-                if (candidate.isCompatibleWith(forkOptions)) {
+                if (candidate.isFailed()) {
+                    emitUnexpectedWorkerFailureWarning(candidate);
                     it.remove();
-                    if (candidate.getLogLevel() != currentLogLevel) {
-                        // TODO: Send a message to workers to change their log level rather than stopping
-                        LOGGER.info("Log level has changed, stopping idle worker daemon with out-of-date log level.");
-                        candidate.stop();
-                    } else {
-                        return candidate;
+                } else {
+                    if (candidate.isCompatibleWith(forkOptions)) {
+                        it.remove();
+                        if (candidate.getLogLevel() != currentLogLevel) {
+                            // TODO: Send a message to workers to change their log level rather than stopping
+                            LOGGER.info("Log level has changed, stopping idle worker daemon with out-of-date log level.");
+                            candidate.stop();
+                        } else {
+                            return candidate;
+                        }
                     }
                 }
             }
             return null;
         }
     }
 
+    private static void emitUnexpectedWorkerFailureWarning(WorkerDaemonClient candidate) {
+        if (candidate.getExitCode().isPresent()) {
+            int exitCode = candidate.getExitCode().get();
+            if (exitCode > 127) {
+                LOGGER.warn("Worker daemon '" + candidate.getDisplayName() + "' exited unexpectedly after being killed with signal " + (exitCode - 128) + ".  This is likely because an external process has killed the worker.");
+            } else {
+                LOGGER.warn("Worker daemon '" + candidate.getDisplayName() + "' exited unexpectedly with exit code " + exitCode + ".");
+            }
+        } else {
+            LOGGER.warn("Worker daemon '" + candidate.getDisplayName() + "' exited unexpectedly.");
```

Comment:
Is there any way we can rework the interface of `WorkerDaemonClient` or something similar to avoid this branch? I don't see this ever actually getting executed at runtime, since we only call this method when `candidate.isFailed()`, which by means `candidate.getExitCode()` will always be present.

Reply:
Good point.  I was thinking it was possible that we might have an ExecResult without an exit code, but that's never the case, so this is pointless.  I've removed it.